src/compiler/crystal/codegen/cache_dir.cr

```diff
@@ -0,0 +1,95 @@
+module Crystal
+  # Manages cache files in the ".crystal" directory.
+  #
+  # For each compiled program a directory is created in the cache
+  # that stores .bc and .o files that could possibly be reused
+  # from a previous compilation.
+  #
+  # To keep the cache dir small, only the 10 most recently used
+  # directories are kept. We use the directory's modification
+  # time for this.
+  class CacheDir
+    def self.instance
+      @@instance ||= new(Config.cache_dir)
+    end
+
+    private def initialize(@dir : String)
+      @initialized = false
+    end
+
+    # Returns the directory where cache files related to the
+    # given sources will be stored. The directory will be
+    # created if it doesn't exist.
+    def directory_for(sources : Array(Compiler::Source))
+      directory_for(sources.first.filename)
+    end
+
+    # Returns the directory where cache files related to the
+    # given filenames will be stored. The directory will be
+    # created if it doesn't exist.
+    def directory_for(filename : String)
+      init
+
+      name = filename.gsub('/', '-')
+      while name.starts_with?('-')
+        name = name[1..-1]
+      end
+      output_dir = File.join(@dir, name)
+      Dir.mkdir_p(output_dir)
+      output_dir
+    end
+
+    # Keeps the 10 most recently used directories in the cache,
+    # and removes all others. This also removes non-directory
+    # files inside the cache directory (temporary executables
+    # resulting from `crystal run` or `run` macro calls).
+    def cleanup
+      init
+
+      entries = gather_cache_entries
+      cleanup_dirs(entries)
+      cleanup_files(entries)
+    end
+
+    # Returns a filename that has prepended the cache directory.
+    def join(filename)
+      init
+      File.join(@dir, filename)
+    end
+
+    private def init
+      return if @initialized
+
+      @initialized = true
+
+      # Check that we can create the cache directory. Otherwise
+      # default to ".crystal", which should probably be writable.
+      begin
+        Dir.mkdir_p(@dir)
+      rescue Errno
+        @dir = ".crystal"
```

Comment:
I will `Dir.mkdir_p(".crystal")`, and if that fails too I'll give a compiler error. Sounds good?